patches/server/1073-Extend-HumanEntity-dropItem-API.patch

```diff
@@ -0,0 +1,95 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Strokkur24 <strokkur24@gmail.com>
+Date: Sat, 30 Nov 2024 11:54:38 +0100
+Subject: [PATCH] Extend HumanEntity#dropItem API
+
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+index e345cdbfab44a0f5da80d738798dbb4424b7ab5c..6a196d49de7b4e592efbbf32089b29dc274ed0f2 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+@@ -801,6 +806,84 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
+         // Paper end - Fix HumanEntity#drop not updating the client inv
+     }
+ 
++    // Paper start - Extend HumanEntity#dropItem API
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final ItemStack itemStack) {
++        return this.dropItem(itemStack, false, false);
++    }
++
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final ItemStack itemStack, final boolean retainOwnership) {
++        return this.dropItem(itemStack, retainOwnership, false);
++    }
++
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final ItemStack itemStack, final boolean retainOwnership, final boolean throwRandomly) {
++        return this.dropItem(this.inventory.first(itemStack), retainOwnership, throwRandomly);
++    }
++
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final int slot) {
++        return this.dropItem(slot, false, false);
++    }
++
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final int slot, final boolean retainOwnership) {
++        return this.dropItem(slot, retainOwnership, false);
++    }
++
++    @Override
++    public @org.jetbrains.annotations.Nullable org.bukkit.entity.Item dropItem(final int slot, final boolean retainOwnership, final boolean throwRandomly) {
++        // Make sure the slot is in bounds
++        if (slot < 0 || this.inventory.getSize() <= slot) {
```

Comment:
If this method stays (see my previous comment) an exception would be more appropriate

Reply:
Also just to note: As I am using this method when dropping via ItemStack, this would also throw an exception each time the ItemStack is not found, which seems counterproductive. So same case as the other one

Reply:
In that case of the slot is outside the inventory bounds we should definitely throw

Reply:
The issue is:
1. I completely agree with you on user input. If you put `37` as the slot, is should throw an `IndexOutOfBoundsException`
2. But if they put in an ItemStack that hasn't been found, an `IndexOutOfBoundsException` feels weird

Reply:
Maybe I can add a check inside the ItemStack overload one to check if the ItemStack exists, and only then pass the slot in?

Reply:
Yeah okay so I made it now that if you pass an incorrect slot in, it throws an `IndexOutOfBoundsException`, like this:
```
[20:14:39 ERROR]: Command exception: /drop -1                                                                                                                                                           
java.lang.IndexOutOfBoundsException: Slot -1 out of range for inventory of size 41                                                                                                                      
        at org.bukkit.craftbukkit.entity.CraftHumanEntity.dropItem(CraftHumanEntity.java:839) ~[main/:?]
        at org.bukkit.craftbukkit.entity.CraftHumanEntity.dropItem(CraftHumanEntity.java:832) ~[main/:?]
        at test-plugin-1.0.0-SNAPSHOT.jar/io.papermc.testplugin.TestPlugin.logic(TestPlugin.java:29) ~[test-plugin-1.0.0-SNAPSHOT.jar:?]
        at test-plugin-1.0.0-SNAPSHOT.jar/io.papermc.testplugin.TestPlugin$1.execute(TestPlugin.java:21) ~[test-plugin-1.0.0-SNAPSHOT.jar:?]
        at io.papermc.paper.command.brigadier.bukkit.BukkitCommandNode$BukkitBrigCommand.run(BukkitCommandNode.java:82) ~[main/:?]
        at com.mojang.brigadier.context.ContextChain.runExecutable(ContextChain.java:73) ~[brigadier-1.3.10.jar:?]
        at net.minecraft.commands.execution.tasks.ExecuteCommand.execute(ExecuteCommand.java:31) ~[filteredMinecraft.jar:?]
        at net.minecraft.commands.execution.tasks.ExecuteCommand.execute(ExecuteCommand.java:19) ~[filteredMinecraft.jar:?]
        at net.minecraft.commands.execution.UnboundEntryAction.lambda$bind$0(UnboundEntryAction.java:8) ~[filteredMinecraft.jar:?]
        at net.minecraft.commands.execution.CommandQueueEntry.execute(CommandQueueEntry.java:5) ~[filteredMinecraft.jar:?]
        at net.minecraft.commands.execution.ExecutionContext.runCommandQueue(ExecutionContext.java:103) ~[filteredMinecraft.jar:?]
        at net.minecraft.commands.Commands.executeCommandInContext(Commands.java:448) ~[main/:?]
        at net.minecraft.commands.Commands.performCommand(Commands.java:355) ~[main/:?]
        at net.minecraft.commands.Commands.performCommand(Commands.java:342) ~[main/:?]
        at net.minecraft.commands.Commands.performCommand(Commands.java:337) ~[main/:?]
        at net.minecraft.server.network.ServerGamePacketListenerImpl.performUnsignedChatCommand(ServerGamePacketListenerImpl.java:2241) ~[main/:?]
        at net.minecraft.server.network.ServerGamePacketListenerImpl.lambda$handleChatCommand$11(ServerGamePacketListenerImpl.java:2215) ~[main/:?]
        at net.minecraft.server.TickTask.run(TickTask.java:18) ~[filteredMinecraft.jar:?]
        at net.minecraft.util.thread.BlockableEventLoop.doRunTask(BlockableEventLoop.java:154) ~[main/:?]
        at net.minecraft.util.thread.ReentrantBlockableEventLoop.doRunTask(ReentrantBlockableEventLoop.java:24) ~[filteredMinecraft.jar:?]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:1577) ~[main/:?]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:207) ~[main/:?]
        at net.minecraft.util.thread.BlockableEventLoop.pollTask(BlockableEventLoop.java:128) ~[main/:?]
        at net.minecraft.server.MinecraftServer.pollTaskInternal(MinecraftServer.java:1554) ~[main/:?]
        at net.minecraft.server.MinecraftServer.pollTask(MinecraftServer.java:1547) ~[main/:?]
        at net.minecraft.util.thread.BlockableEventLoop.managedBlock(BlockableEventLoop.java:138) ~[main/:?]
        at net.minecraft.server.MinecraftServer.managedBlock(MinecraftServer.java:1498) ~[main/:?]
        at net.minecraft.server.MinecraftServer.waitUntilNextTick(MinecraftServer.java:1508) ~[main/:?]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1344) ~[main/:?]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:340) ~[main/:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
```